torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,159 @@ namespace rpc {
+  if (!rpcAgentRunning_.exchange(false)) {
+    return;
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+  shutdownImpl();
+}
+
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  TORCH_CHECK(maxRetries >= 0, "maxRetries cannot be negative.");
+  auto originalFuture = std::make_shared<FutureMessage>();
+  steady_clock_time_point newTime = computeNewRpcRetryTime(/* retryCount */ 0);
+  Message& msg = message;
+  auto fm = send(to, std::move(message));
+  auto firstRetryRpc = std::make_shared<RpcRetryInfo>(
+      to, std::move(msg), originalFuture, /* retryCount */ 0, maxRetries);
+
+  fm->addCallback([this, newTime, firstRetryRpc](
+                      const rpc::Message& lambdaMessage,
+                      const c10::optional<utils::FutureError>& futErr) {
+    rpcRetryCallback(lambdaMessage, futErr, newTime, firstRetryRpc);
+  });
+
+  return originalFuture;
+}
+
+void RpcAgent::retryExpiredRpcs() {
+  while (rpcAgentRunning_.load()) {
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
+
+    // We must continue sleeping as long as the RPC Agent is running and when
+    // either the Retry Map is empty, or when the Retry Map's earliest expiring
+    // RPC is set to be retried in the future.
+    if (rpcRetryMap_.empty()) {
+      rpcRetryMapCV_.wait(lock, [&, this] {
+        return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+      });
+    }
+
+    steady_clock_time_point earliestTimeout;
 
-RpcAgent::~RpcAgent() = default;
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
+        return !(rpcAgentRunning_.load());
+      });
+    }
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+
+    // Updating these since something may have been added to the map while this
+    // thread was sleeping.
+    earliestTimeout = rpcRetryMap_.begin()->first;
+    auto& earliestRpcList = rpcRetryMap_.begin()->second;
+
+    // We iterate through all the RPC's set to be retried at the current
+    // timepoint, resend those RPC's, and add a callback that either schedules
+    // the RPC for a future retry or marks it with success/error depending on
+    // the outcome of the current send. Then, we clean up the rpcRetryMap_.
+    for (auto it = earliestRpcList.begin(); it != earliestRpcList.end();
+         /* no increment */) {
+      auto& earliestRpc = *it;
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
+      earliestRpc->retryCount_++;
+
+      fm->addCallback([this, newTime, earliestRpc](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        rpcRetryCallback(message, futErr, newTime, earliestRpc);
```

Comment:
Don't hold the lock across this addCallback() (I'd also say that about send(), but slightly less critical there).
This is inviting a deadlock on an insta-fail.
Typically, you make a list, then do after releasing the lock.

Reply:
good catch! The callback here might launch right away if the fm has been marked complete, which would lead to deadlock as it will try to acquire `rpcRetryMutex_` again.